src/core/config/srp_client.h

```diff
@@ -0,0 +1,196 @@
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_SRP_CLIENT_EARLY_LEASE_RENEW_FACTOR_NUMERATOR
+ *
+ * Specifies the numerator of early lease renewal factor.
+ *
+ * This value is used for opportunistic early refresh behave. When sending an SRP update, the services that are not yet
+ * expired but are close, are allowed to refresh early and are included in the SRP update.
+ *
+ * The "early lease renewal interval" is used to determine if a service can renew early. The interval is calculated by
+ * multiplying the accepted lease interval by the "early lease renewal factor" which is given as a fraction (numerator
+ * and denominator).
+ *
+ * If the factor is set to zero (numerator=0, denominator=1), the opportunistic early refresh behavior is disabled.
+ * If  denominator is set to zero (the factor is set to infinity), then all services (including previously registered
+ * ones) are always included in SRP update message.
+ *
+ * Default value is 1/2 (i.e., services that are within half of the lease interval are allowed to refresh early).
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_SRP_CLIENT_EARLY_LEASE_RENEW_FACTOR_NUMERATOR
+#define OPENTHREAD_CONFIG_SRP_CLIENT_EARLY_LEASE_RENEW_FACTOR_NUMERATOR 1
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_SRP_CLIENT_EARLY_LEASE_RENEW_FACTOR_DENOMINATOR
+ *
+ * Specifies the denominator of early lease renewal factor.
+ *
+ * Please see OPENTHREAD_CONFIG_SRP_CLIENT_EARLY_LEASE_RENEW_FACTOR_NUMERATOR for more details.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_SRP_CLIENT_EARLY_LEASE_RENEW_FACTOR_DENOMINATOR
+#define OPENTHREAD_CONFIG_SRP_CLIENT_EARLY_LEASE_RENEW_FACTOR_DENOMINATOR 2
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_SRP_CLIENT_UPDATE_TX_DELAY
+ *
+ * Specifies the (short) delay (in msec) after an update is required before SRP client sends the update message.
+ *
+ * When there is a change (e.g., a new service is added/removed) that requires an update, the SRP client will wait for
+ * a short delay before preparing and sending an SRP update message to server. This allows user to provide more change
+ * that are then all sent in same update message. The delay is only applied on the first change that triggers an
+ * update message transmission. Subsequent changes (API calls) while waiting for the tx to start will not reset the
+ * delay timer.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_SRP_CLIENT_UPDATE_TX_DELAY
+#define OPENTHREAD_CONFIG_SRP_CLIENT_UPDATE_TX_DELAY 10
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_SRP_CLIENT_MIN_RETRY_WAIT_INTERVAL
+ *
+ * Specifies the minimum wait interval (in msec) between SRP update message retries.
+ *
+ * The update message is retransmitted if there is no response from server or if server rejects the update. The wait
+ * interval starts from the minimum value and is increased by the growth factor every failure up to the max value.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_SRP_CLIENT_MIN_RETRY_WAIT_INTERVAL
+#define OPENTHREAD_CONFIG_SRP_CLIENT_MIN_RETRY_WAIT_INTERVAL 800
```

Comment:
This value may be too small if there are Advertising Proxy on the server side. To detect name conflicts as possible, implementations may wait for the Advertising Proxy finishing [mDNS Probing](https://tools.ietf.org/html/rfc6762#section-8.1) before responding to the client. The maximum delay for mDNS probing is about `250ms * 4`.

I think a reasonable value could be `(1000 + 800)`. Thoughts?

Reply:
Good point! Let's go for 1500ms as the minimum. I will also change the growth factor from 2 to 1.8. So that we don't increase the delay too quickly and have more initial attempts.

The delays would be: 1.5, 2.7, 4.9, 8.7, 15.8, 28.3, 51.0, ... And we will have 5 attempts in first ~33 second or so.

Reply:
Actually we need to take jitter into account as well. 

Here is the numbers I went for.
- MIN_RETRY = 1800ms
- JITTER = 500ms (this way the first wait would be > 1300ms even with min jitter)
- GROWTH FACTOR = 1.7

With this the attempts are at 1.8, 3.1, 5.2, 8.8, 15.0, 25.6, 43..4, 73.9, ...

Reply:
Thanks, SGTM!